pkg/util/cephcmds.go

```diff
@@ -0,0 +1,364 @@
+		return err
+	}
+
+	return nil
+}
+
+// ErrKeyNotFound is returned when requested key in omap is not found
+type ErrKeyNotFound struct {
+	keyName string
+	err     error
+}
+
+func (e ErrKeyNotFound) Error() string {
+	return e.err.Error()
+}
+
+// GetOMapValue gets the value for the given key from the named omap
+func GetOMapValue(monitors, adminID, key, poolName, oMapName, oMapKey string) (string, error) {
+	// Command: "rados <options> getomapval oMapName oMapKey <outfile>"
+	// No such key: replicapool/csi.volumes.directory.default/csi.volname
+	tmpFile, err := ioutil.TempFile("", "omap-get-")
+	if err != nil {
+		klog.Errorf("failed creating a temporary file for key contents")
+		return "", err
+	}
+	defer tmpFile.Close()
+	defer os.Remove(tmpFile.Name())
+
+	stdout, _, err := execCommand(
+		"rados",
+		"-m", monitors,
+		"--id", adminID,
+		"--key="+key,
+		"-c", CephConfigPath,
+		"-p", poolName,
+		"getomapval", oMapName, oMapKey, tmpFile.Name())
+	if err != nil {
+		// no logs, as attempting to check for key/value is done even on regular call sequences
+		if strings.Contains(string(stdout), "No such key: "+poolName+"/"+oMapName+"/"+oMapKey) {
+			return "", ErrKeyNotFound{poolName + "/" + oMapName + "/" + oMapKey, err}
+		}
+		return "", err
+	}
+
+	keyValue, err := ioutil.ReadAll(tmpFile)
+	return string(keyValue), err
+}
+
+// RemoveOMapKey removes the omap key from the given omap name
+func RemoveOMapKey(monitors, adminID, key, poolName, oMapName, oMapKey string) error {
+	// Command: "rados <options> rmomapkey oMapName oMapKey"
+
+	_, _, err := execCommand(
+		"rados",
+		"-m", monitors,
+		"--id", adminID,
+		"--key="+key,
+		"-c", CephConfigPath,
+		"-p", poolName,
+		"rmomapkey", oMapName, oMapKey)
+	if err != nil {
+		// NOTE: Missing omap key removal does not return an error
+		klog.Errorf("failed removing key (%s), from omap (%s) in "+
+			"pool (%s): (%v)", oMapKey, oMapName, poolName, err)
+		return err
+	}
+
+	return nil
+}
+
+// ErrOMapNotFound is returned when named omap is not found in rados
+type ErrOMapNotFound struct {
+	oMapName string
+	err      error
+}
+
+func (e ErrOMapNotFound) Error() string {
+	return e.err.Error()
+}
+
+// RemoveOMap removes the entire omap name passed in and returns ErrOMapNotFound is provided omap
+// is not found in rados
+func RemoveOMap(monitors, adminID, key, poolName, oMapName string) error {
+	// Command: "rados <options> rm oMapName"
+
+	stdout, _, err := execCommand(
+		"rados",
+		"-m", monitors,
+		"--id", adminID,
+		"--key="+key,
+		"-c", CephConfigPath,
+		"-p", poolName,
+		"rm", oMapName)
+	if err != nil {
+		klog.Errorf("failed removing omap (%s) in pool (%s): (%v)", oMapName, poolName, err)
+		if strings.Contains(string(stdout), "error removing "+poolName+">"+oMapName+
+			": (2) No such file or directory") {
+			return ErrOMapNotFound{oMapName, err}
+		}
+		return err
+	}
+
+	return nil
+}
+
+// RBDImageInfo strongly typed JSON spec for image info
```

Comment:
Nit: these remaining things seem like they belong in the ```rbd``` package

Reply:
I assume this is for the 2 functions `GetSnapInfo` and `GetImageInfo`, would that be right?

My reasoning is that these functions run a Ceph/rbd command and return the results, hence should belong here.

Here are more elaborate reasons/thinking about cephcmds.go and rbd package in general that I have,
1) rbd package functions operate on rbdVol or rbdSnap structures
- Some functions inside rbd package also invoke a corresponding Ceph command (pretty much functions in the file rbd_util.go)
  - This is done based on extracting required information from the rbd[Vol|Snap] structure, but makes the function unusable outside the rbd_util.go context
- Unit testing the rbd_util functions that invoke Ceph commands is not easily feasible

2) cephcmds.go aims to be a pure "invoke said Ceph command" series of functions, and is not tied to rbdVol|Snap structures
- makes these functions in cephcmds.go reusable (when required) and also independently testable
- In the future if any call output/input/parsing needs to change, it can be handled in one place (cephcmds.go)

- Initially was written to avoid constantly extracting required information from rbdVol|Sbap structures (calling getRBDKey in each function say) to make Ceph calls, like in the function reserveRBDVol
- but, over time in the PR made it a wrapper over Ceph commands, in the hope that in the future we would be able to test and update this independently

Ideally I would suggest that we move a bulk of the rbd_util.go code that invokes Ceph commands to this file, and as a result would prefer to retain the following 2 functions here.

Thoughts?

Reply:
I wasn't concerned about the generic ```ceph``` CLI commands (beyond the usual brittleness of relying on CLI commands vice a testable API). The mixing of RBD-unique data types, to me, is polluting the diving line between the two. Perhaps the ```rbd``` package just needs an equivalent ```rbdcmds``` file?

Reply:
Not attempting to press a point, as I do not have any objections to move the functions to the rbd_util.go file (as on further thinking, testing the functions in either package should be feasible and nothing special), BUT,
If the structures were named ImageInfo or some such and not RBD prefixed would that have changed your mind?

I ask as the RBD prefix naming of these was more from the rbd CLI usage and return values, than the rbd golang package usage in this repository, and hence was curious.

Reply:
Moved rbd specifc CLI calls to the rbd package.